src/lib/packager/clients/pkg_finish.rb

```diff
@@ -0,0 +1,171 @@
+require "packages/repository"
+
+module Yast
+  class PkgFinishClient < ::Installation::FinishClient
+    include Yast::I18n
+    include Yast::Logger
+
+    # Path to libzypp repositories
+    REPOS_DIR = "/etc/zypp/repos.d"
+    # Repository schemes to disable (disable_local_repos)
+    SCHEMES_TO_DISABLE = [:cd, :dvd]
+    # Path to failed_packages file
+    FAILED_PACKAGES_PATH = "/var/lib/YaST2/failed_packages"
+    # Command to create a tar.gz to back-up old repositories
+    TAR_CMD = "mkdir -p '%<source>s' && cd '%<source>s' "\
+      "&& /bin/tar -czf '%<archive>s' '%<target>s'"
+
+
+    # Constructor
+    def initialize
+      textdomain "packager"
+
+      Yast.import "Pkg"
+      Yast.import "Installation"
+      Yast.import "Mode"
+      Yast.import "Stage"
+      Yast.import "String"
+      Yast.import "FileUtils"
+      Yast.import "Packages"
+    end
+
+    # @see Implements ::Installation::FinishClient#modes
+    def modes
+      [:installation, :update, :autoinst]
+    end
+
+    # @see Implements ::Installation::FinishClient#title
+    def title
+      _("Saving the software manager configuration...")
+    end
+
+    # @see Implements ::Installation::FinishClient#write
+    def write
+      # Remove (backup) all sources not used during the update
+      # BNC #556469: Backup and remove all the old repositories before any Pkg::SourceSaveAll call
+      BackUpAllTargetSources() if Stage.initial && Mode.update
+
+      # See bnc #384827, #381360
+      if Mode.update
+        log.info("Adding default repositories")
+        WFM.call("inst_extrasources")
+      end
+
+      disable_uneeded_repos(SCHEMES_TO_DISABLE)
+
+      # save all repositories and finish target
+      Pkg.SourceSaveAll
+      Pkg.TargetFinish
+
+      # save repository metadata cache to the installed system
+      # (needs to be done _after_ saving repositories, see bnc#700881)
+      Pkg.SourceCacheCopyTo(Installation.destdir)
+
+      # copy list of failed packages to installed system
+      WFM.Execute(
+        path(".local.bash"),
+        format("test -f %<path>s && /bin/cp -a %<path>s '%<destdir>s%<path>s'",
+          path: FAILED_PACKAGES_PATH, destdir: String.Quote(Installation.destdir)))
+    end
+
+  private
+
+    # During upgrade, old sources are reinitialized
+    # either as enabled or disabled.
+    # The old sources from targed should go away.
+    def BackUpAllTargetSources
+      Yast.import "Directory"
+
+      if !File.exist?(REPOS_DIR)
+        log.error("Directory #{REPOS_DIR} doesn't exist!")
+        return
+      end
+
+      current_repos = SCR.Read(path(".target.dir"), REPOS_DIR)
+
+      if current_repos.nil? || current_repos.empty?
+        log.warn("There are currently no repos in #{REPOS_DIR} conf dir")
+        return
+      else
+        log.info("These repos currently exist on a target: #{current_repos}")
+      end
+
+      cmd = WFM.Execute(path(".local.bash_output"), "date +%Y%m%d-%H%M%S")
```

Comment:
[Time.strftime](http://ruby-doc.org/core-2.2.0/Time.html#method-i-strftime) would be nice here, but it seems that lot of places in this client could be Rubyfied... So maybe at least few of them could be improved... :wink: